src/app/public/modules/wait/wait-adapter.service.ts

```diff
@@ -21,13 +36,69 @@ export class SkyWaitAdapterService {
     waitEl: ElementRef,
     isFullPage: boolean,
     isWaiting: boolean,
+    id: string,
     isNonBlocking = false
-  ): void {
+  ) {
     let busyEl = isFullPage ? document.body : waitEl.nativeElement.parentElement;
     let state = isWaiting ? 'true' : undefined;
 
     if (!isNonBlocking) {
       this.renderer.setElementAttribute(busyEl, 'aria-busy', state);
+
+      if (isWaiting) {
+        // Propagate tab navigation if attempted into waited element
+        if (!isFullPage) {
+          let focusListenerFunc = this.renderer.listen(this.windowRef.getWindow(), 'keyup', (event: any) => {
+            if (event.key.toLowerCase() === 'tab' && busyEl.contains(document.activeElement)) {
+              this.focusNextElement(busyEl, event.shiftKey);
+            }
+          });
+          this.parentListeners[id] = focusListenerFunc;
+        } else {
+          // Prevent tab navigation within the waited page
+          let endListenerFunc = this.renderer.listen(busyEl, 'keydown', (event: KeyboardEvent) => {
+            if (event.key.toLowerCase() === 'tab') {
+              event.preventDefault();
+            }
+          });
+          this.parentListeners[id] = endListenerFunc;
+        }
+      } else if (id in this.parentListeners) {
+        // Clean up existing listener
+        this.parentListeners[id]();
+        delete this.parentListeners[id];
+      }
+    }
+  }
+
+  private focusNextElement(parentElement: any, shiftKey: boolean): void {
+    // Select all possible focussable elements
+    let focussableElements =
+      'a[href], ' +
+      'area[href], ' +
+      'input:not([disabled]):not([tabindex=\'-1\']), ' +
+      'button:not([disabled]):not([tabindex=\'-1\']), ' +
+      'select:not([disabled]):not([tabindex=\'-1\']), ' +
+      'textarea:not([disabled]):not([tabindex=\'-1\']), ' +
+      'iframe, object, embed, ' +
+      '*[tabindex]:not([tabindex=\'-1\']), ' +
+      '*[contenteditable=true]';
+     let focussable = Array.prototype.filter.call(document.body.querySelectorAll(focussableElements),
+      (element: any) => {
+        return element.offsetWidth > 0 || element.offsetHeight > 0 || element === document.activeElement;
+      });
+     // If shift tab, go in the other direction
+    let modifier = 1;
+    if (shiftKey) {
+      modifier = -1;
+    }
+     // Find the next navigable element that isn't waiting
+    let curIndex = focussable.indexOf(document.activeElement) + modifier;
+    while (!focussable[curIndex] || parentElement.contains(focussable[curIndex]) || parentElement === focussable[curIndex]) {
```

Comment:
I'm not sure if this `while` loop is the cause, but if I put an `<a>` element in the white div of the visual test, and try to tab into it while the wait is triggered, it crashes my browser. This happens in both FF and Chrome :(

Reply:
Yeah, it was the cause. I fixed its bounds. I'm confused as to how it was working before though :/